### PR TITLE
WIP Denormalize entity subtypes (pr2)

### LINKED ISSUE #11134: master: Seed needs to create user profile fields as annotations
I'm working on this in #11132

### PR DESCRIPTION
**Pull request by mrclay**
ongoing work on #11109

- [ ] Refuse schema migration if subtypes table holds `user`, `group` or `site` subtype
- [ ] Check for objects without subtype
- [ ] Document query conventions: if `type` and `subtype` are set, the will also be search for literally (`true`, `false`, `0` will be case to string in in `e.subtype IN ()` query); if not set, all entities of type and subtype will be returned
- [ ] `register_user` should accept subtype argument
- [ ] verify it fixes #11134

### CODE REVIEW

File: engine/schema/migrations/20170804152152_align_subtype_columns.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		];
+
+		$table = $this->table('river');
+		$table->changeColumn('subtype', 'string', $options)->save();
+
+		$table = $this->table('entities');
+		$table->changeColumn('subtype', 'string', $options)->save();
+
+		$table = $this->table('system_log');
+		$table->changeColumn('object_subtype', 'string', $options)->save();
```
Comment (mrclay) [inline engine/schema/migrations/20170804152152_align_subtype_columns.php]: Upgrade fails here with `1071 Specified key was too long; max key length is 1000 bytes`. I don't get why.

**Comment by hypeJunction** [comment]
I think we should just keep avoid tables with utf8mb4 encoding if we have utf8 columns

File: engine/schema/migrations/20170804152152_align_subtype_columns.php (hunk 1 above)
Comment (hypeJunction) [inline engine/schema/migrations/20170804152152_align_subtype_columns.php]: I think that table has a composite key. Maybe remove and create it with a limit

**Comment by hypeJunction** [comment]
https://github.com/Elgg/Elgg/pull/11131 had a similar problem

**Comment by hypeJunction** [comment]
Oh, you got the tests to pass! 😎

**Comment by hypeJunction** [comment]
Hm, something is weird with simpletests suite. I saw this before. Exception thrown but test exits with success.

**Comment by mrclay** [comment]
We run migrations which change the schema but don't update the DB version until the real upgrades have run. Hence, Elgg\Database\ConfigTable::set ends up trying to write a site_guid column.

**Comment by hypeJunction** [comment]
Let's roll in a simpler version of #11131. I will see if we can move utf8mb4 change into an asynchronous upgrade

**Comment by hypeJunction** [comment]
You should be able to work on this now without problems with previous migrations and config.

**Comment by hypeJunction** [comment]
Rebased version in #11310 with most to-dos checked.

**Comment by jeabakker** [comment]
@hypeJunction so this is replaced by #11310?

**Comment by hypeJunction** [comment]
Yes. You can close it.

**Comment by jdalsem** [comment]
@hypeJunction did you also cover these todo's in your PR?
> - register_user should accept subtype argument
> - verify it fixes #11134

**Comment by hypeJunction** [comment]
> register_user should accept subtype argument

That's already done in one of the previous PRs

> verify it fixes #11134

I don't care about #11134 in this PR. It has nothing to do with subtypes.

**Pull request closed without merging**